### ISSUE TITLE
"runners.actions.summerwind.dev" is invalid: metadata.annotations: Too long: must have at most 262144 bytes

### ISSUE THREAD

**Issue by ealasgarov** [actions/actions-runner-controller#1317]
**Describe the bug**

When I'm installing actions-runner-controller with:

"kubectl apply -f https://github.com/actions-runner-controller/actions-runner-controller/releases/download/v0.22.2/actions-runner-controller.yaml" command

I am getting below error:

"Error from server (Invalid): error when creating "https://github.com/actions-runner-controller/actions-runner-controller/releases/download/v0.22.2/actions-runner-controller.yaml": CustomResourceDefinition.apiextensions.k8s.io "runners.actions.summerwind.dev" is invalid: metadata.annotations: Too long: must have at most 262144 bytes"

**To Reproduce**

1. Run  'kubectl apply -f https://github.com/actions-runner-controller/actions-runner-controller/releases/download/v0.22.2/actions-runner-controller.yaml'

**Expected behavior**

Resources should be installed as described in the Readme

**Screenshots**

![image](https://user-images.githubusercontent.com/86415933/162255220-be17c898-a2af-420e-aa0b-098d6aa2a684.png)

**Environment (please complete the following information):**

- [x] actions-runner-controller 0.22.2 (latest)
- [ ] AKS (kubernetes) version 1.22.6 (latest stable version)
- [ ] cert-manager installed, latest version

**Additional context**

I have downloaded that **actions-runner-controller.yaml** and I can see that custom resource definition has more than 30k lines, is that right?

**Comment by mumoshu** [comment]
@ealasgarov Hey. This is a known issue in Kubernetes(and kubectl-apply) that results in that an annotation automatically generated for server-side apply results in exceeding the K8s' own limit. Try using `kubectl create` instead.

**Comment by KENNYSOFT** [comment]
Possibly duplicate of #1102

**Comment by ealasgarov** [comment]
thanks a lot for comments, will try that!

**Comment by billimek** [comment]
> can we maybe change in the readme the `kubectl apply` by `kubectl create` just to avoid having to search for that?

@jonathanbeber do you happen to know if the `kubectl create` approach will work for _updating_ CRDs?  

I'm thinking about a scenario where the CRDs are already installed and we are upgrading to a new version which has changes to the CRDs.  From my experience, `kubectl create` will fail if the resource already exists.  If this is the case, it would require a deletion of the CRDs first followed by a crash `create`. 

The consequence of deleting the CRDs is that all the `RunnerDeployments`, `Runners`, etc will also be subsequently removed unless we maybe also add the `--cascade=orphan` argument to the `kubectl delete` operation?

To your point, I think we need to update the docs given the size of the CRDs and `apply` no longer working.

**Comment by mumoshu** [comment]
@billimek In case you missed it, you can use `kubectl replace` for upgrading CRDs as documented in https://github.com/actions-runner-controller/actions-runner-controller/blob/master/charts/actions-runner-controller/docs/UPGRADING.md#steps

**Comment by zerodayyy** [comment]
Are there any solutions to work around this issue when using ArgoCD? Unfortunately it's not possible to change the underlying logic there and installing CRDs manually is not an option

**Comment by toast-gear** [comment]
I don't think so outside of keeping custom version of the chart and adding the the `argocd.argoproj.io/sync-options: Replace=true` annotation to the CRDs in your custom chart

**Comment by mumoshu** [comment]
Also see #1391 and https://github.com/actions-runner-controller/actions-runner-controller/issues/1256#issuecomment-1082361032, https://github.com/actions-runner-controller/actions-runner-controller/issues/1102#issuecomment-1041138914

**Comment by mumoshu** [comment]
To me https://github.com/prometheus-operator/prometheus-operator/issues/4439#issuecomment-1030198014 looks like the best way; TL;DR; you can set the sync option via ArgoCD Application spec.

**Comment by shadiramadan** [comment]
I do think it would help to distribute the CRDs separately.

**Comment by SebastianScherer88** [comment]
I had the same issue and can confirm that the [workaround shown here](https://www.pcbaecker.com/posts/kubernetes-crd-annotation-too-long-fix/):

```
kubectl create -f ...  // For creating the resource
kubectl replace -f ... // For updating the resource
```

worked for me

**Comment by yyvess** [comment]
Try the option server-side

_kubectl apply -f xxx.yaml --server-side_

https://kubernetes.io/docs/reference/using-api/server-side-apply/

**Comment by brnpimentel** [comment]
> Try the option server-side
> 
> _kubectl apply -f xxx.yaml --server-side_
> 
> https://kubernetes.io/docs/reference/using-api/server-side-apply/

`Error: unknown flag: --server-side`

**Comment by jspc** [comment]
> Try the option server-side
> 
> _kubectl apply -f xxx.yaml --server-side_
> 
> https://kubernetes.io/docs/reference/using-api/server-side-apply/

This worked perfectly for me. In my particular use case, I was running a github actions job that was just running the line:

```
$ kubectl apply -f https://github.com/actions/actions-runner-controller/releases/download/v0.22.0/actions-runner-controller.yaml
```

And so following the instruction to use `kubctl create` and then `kubectl update` would have sucked, frankly, because I'd have had to put some logic in to check whether a/some/all the CRDs existed first to determine which command to run.

The `--server-side` flag is what got my over that hurdle.

A note, though; if you've already tried to install the CRDs via `kubectl apply -f...` in the past, you will see a series of errors about ownership of various types.

Personally, I just deleted them (`kubectl delete -f https://github.com/actions/actions-runner-controller/releases/download/v0.22.0/actions-runner-controller.yaml`) myself, then ran again with the `--server-side` flag, which worked perfectly.

Another approach is to run `kubectl apply -f ...` with the flag `--force-conflicts` too, but I wasn't massively in love with doing that.

YMMV

**Comment by mumoshu** [comment]
Thank you everyone for reporting! I can confirm that `kubectl apply --server-side` should be the best way nowadays, as you kindly shared!

**Comment by biclighter81** [comment]
Argocd users can follow this documentation entry: https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#server-side-apply

This syncPolicy works for me:
```
syncPolicy:
    syncOptions:
      - CreateNamespace=true
      - ServerSideApply=true
 ```
